php_phongo.c

```diff
@@ -1963,44 +1975,69 @@ static mongoc_ssl_opt_t* php_phongo_make_ssl_opt(zval* zoptions TSRMLS_DC)
 
 	ssl_opt = ecalloc(1, sizeof(mongoc_ssl_opt_t));
 
+	/* If SSL options are set in the URL, we need to read them and set them on
+	 * the options struct so we can merge potential options from passed in
+	 * driverOptions (zoptions) */
+	if (mongoc_uri_get_ssl(uri)) {
+		php_phongo_mongoc_ssl_opts_from_uri(ssl_opt, uri);
+		any_ssl_option_set = true;
+	}
+
 	/* Check canonical option names first and fall back to SSL context options
 	 * for backwards compatibility. */
 	if (php_array_existsc(zoptions, "allow_invalid_hostname")) {
 		ssl_opt->allow_invalid_hostname = php_array_fetchc_bool(zoptions, "allow_invalid_hostname");
+		any_ssl_option_set              = true;
 	}
 
 	if (php_array_existsc(zoptions, "weak_cert_validation")) {
 		ssl_opt->weak_cert_validation = php_array_fetchc_bool(zoptions, "weak_cert_validation");
+		any_ssl_option_set            = true;
 	} else if (php_array_existsc(zoptions, "allow_self_signed")) {
 		ssl_opt->weak_cert_validation = php_array_fetchc_bool(zoptions, "allow_self_signed");
+		any_ssl_option_set            = true;
 	}
 
 	if (php_array_existsc(zoptions, "pem_file")) {
-		ssl_opt->pem_file = php_phongo_fetch_ssl_opt_string(zoptions, ZEND_STRL("pem_file"));
+		ssl_opt->pem_file  = php_phongo_fetch_ssl_opt_string(zoptions, ZEND_STRL("pem_file"));
```

Comment:
I think there is a risk of leaking memory for any strings that `php_phongo_mongoc_ssl_opts_from_uri()` previously copied with `estrdup()`. Before assigning any strings on the `mongoc_ssl_opt_t` struct after the call to `php_phongo_mongoc_ssl_opts_from_uri()`, you'll need to check if the string is non-null and then `efree()` it.

I might suggest just creating a temporary macro before the first time you need to do this and then undefining it before the end of this function. We do that elsewhere for temporary macros in BulkWrite.c and Query.c.

Before fixing this, it'd be wise to add a test that demonstrates the memory leak. It actually shouldn't require a live server, as we're just building the libmongoc client. You should use `skip_if_not_libmongoc_ssl()`, though.

Reply:
Iv'e updated the code, and added a test file (``tests/manager/manager-set-uri-options-003.phpt``)